### PR TITLE
LogStoreTableFactory should not implements DynamicTableFactory 

### LINKED ISSUE #705: [Bug] LogStoreTableFactory should not implements DynamicTableFactory
### Search before asking

- [X] I searched in the [issues](https://github.com/apache/incubator-paimon/issues) and found nothing similar.

### Paimon version

0.4

### Compute Engine

flink

### Minimal reproduce step

kafka connector sink to paimon table

### What doesn't meet your expectations?

Now, KafkaLogStoreFactory is very easy to conflict with flink kafka connector.
we should not use flink DynamicTableFactory.

### Anything else?

_No response_

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by Alibaba-HZY**
Purpose
LogStoreTableFactory should not implements DynamicTableFactory.if there is no kafka jar in flink/lib, the kafka identifier will find KafkaLogStoreFactory in paimon, and throw KafkaLogStoreFactory is not a DynamicTableSourceFactory exception, which is very confused.so we define the PaimonFactory ,LogStoreTableFactory implements it.
if there is no kafka jar in flink/lib, it will throw org.apache.flink.table.api.ValidationException: Could not find any factory for identifier 'kafka' that implements 'org.apache.flink.table.factories.DynamicTableFactory' in the classpath
 Fix #705

Tests
LogSystemITCase

API and Format
Documentation

### CODE REVIEW

File: docs/content/concepts/external-log-systems.md [hunk 1]
```diff
@@ -42,6 +42,23 @@ If `'log.consistency' = 'eventual'` is set, in order to achieve correct results,
...
+
+Paimon currently supports Flink 1.16, 1.15 and 1.14. We recommend the latest Flink version for a better experience.
+
+Download the flink-sql-connector-kafka jar file with corresponding version.
+
+{{< stable >}}
+
+| Version | Jar                                                                                                                                                             |
+|---|-----------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| Flink 1.16 | [flink-connector-kafka-1.16.1.jar](https://repo.maven.apache.org/maven2/org/apache/flink/flink-sql-connector-kafka/1.16.1/flink-sql-connector-kafka-1.16.0.jar) |
```
Comment (JingsongLi) [inline docs/content/concepts/external-log-systems.md:55]: No `-sql-`?

**Review by JingsongLi — COMMENTED** [review]
Thanks @Alibaba-HZY for the contribution.

I think we need to do a major surgery to delete lots of useless codes.
We just use a very simple factory and its util.

File: docs/content/concepts/external-log-systems.md (hunk 1 above)
Comment (Alibaba-HZY) [reply docs/content/concepts/external-log-systems.md:55]: I left -sql-, i will make it up

**Comment by Alibaba-HZY** [comment]
> I think we need to do a major surgery to delete lots of useless codes.
> We just use a very simple factory and its util.

ok i will delete useless codes

**Comment by Alibaba-HZY** [comment]
> Thanks @Alibaba-HZY for the contribution.
> 
> I think we need to do a major surgery to delete lots of useless codes. We just use a very simple factory and its util.

PaimonFactoryUtil only contains discoverPaimonFactory to find some LogStoreTableFactory implements PaimonFactory.
FlinkFactoryUtil is copied from org.apache.flink.table.factories.FactoryUtil  to find DecodingFormat and EncodingFormat  because we use PaimonFactory so we need rewrite FactoryHelper and FactoryUtil

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/factories/DynamicTablePaimonFactory.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+     * {@link SupportsFilterPushDown}.
+     *
+     * <p>For example, given a database connector, if an option defines the connection timeout,
+     * changing this value does not affect the pipeline topology and can be allowed. However, an
+     * option that defines whether the connector supports {@link SupportsReadingMetadata} or not is
+     * not allowed. The planner might not react to changed abilities anymore.
+     *
+     * @see FormatFactory#forwardOptions()
+     */
+    default Set<ConfigOption<?>> forwardOptions() {
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/factories/DynamicTablePaimonFactory.java:55]: Is this useful?

Comment (Alibaba-HZY) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/factories/DynamicTablePaimonFactory.java]: no  i have remove it

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/log/LogStoreTableFactory.java [hunk 3]
```diff
@@ -50,7 +52,7 @@
  * <p>Log tables are for processing only unbounded data. Support streaming reading and streaming
  * writing.
  */
-public interface LogStoreTableFactory extends DynamicTableFactory {
+public interface LogStoreTableFactory extends DynamicTablePaimonFactory {
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/log/LogStoreTableFactory.java:55]: Maybe we don't need `DynamicTablePaimonFactory` and `PaimonFactory`? Just a `LogStoreTableFactory` with `String factoryIdentifier()` is OK?

Comment (Alibaba-HZY) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/log/LogStoreTableFactory.java]: yes we only need  String factoryIdentifier() ;

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/factories/FlinkFactoryUtil.java [hunk 4]
```diff
@@ -0,0 +1,544 @@
...
+/** Utility for working with {@link Factory}s. */
+public final class FlinkFactoryUtil {
+
+    private static final Logger LOG = LoggerFactory.getLogger(FlinkFactoryUtil.class);
+
+    /**
+     * Describes the property version. This can be used for backwards compatibility in case the
+     * property format changes.
+     */
+    public static final ConfigOption<Integer> PROPERTY_VERSION =
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/factories/FlinkFactoryUtil.java:68]: Useless

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/factories/FlinkFactoryUtil.java [hunk 5]
```diff
@@ -0,0 +1,544 @@
...
+            ConfigOptions.key("property-version")
+                    .intType()
+                    .defaultValue(1)
+                    .withDescription(
+                            "Version of the overall property design. This option is meant for future backwards compatibility.");
+
+    public static final ConfigOption<String> CONNECTOR =
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/factories/FlinkFactoryUtil.java:75]: Useless

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/factories/FlinkFactoryUtil.java [hunk 6]
```diff
@@ -0,0 +1,544 @@
...
+     *
+     * keyFormat = helper.discoverDecodingFormat(DeserializationFormatFactory.class, KEY_FORMAT);
+     * valueFormat = helper.discoverDecodingFormat(DeserializationFormatFactory.class, VALUE_FORMAT);
+     *
+     * helper.validate();
+     *
+     * ... // construct connector with discovered formats
+     * }</pre>
+     *
+     * <p>Note: The format option parameter of {@link
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/factories/FlinkFactoryUtil.java:119]: We don't need such complicated documentation.

**Review by JingsongLi — COMMENTED** [review]
Thanks for the update.
I think our codes have a lot of optimization space.
We should simplify more.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/factories/FlinkFactoryUtil.java (hunk 6 above)
Comment (Alibaba-HZY) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/factories/FlinkFactoryUtil.java]: i have removd

File: docs/content/concepts/external-log-systems.md [hunk 7]
```diff
@@ -42,6 +42,23 @@ If `'log.consistency' = 'eventual'` is set, in order to achieve correct results,
 
 ### Kafka
 
+#### Preparing flink-sql-connector-kafka Jar File
```
Comment (golden-yang) [inline docs/content/concepts/external-log-systems.md:45]: Very timely addition, thanks

**Review by JingsongLi — APPROVED** [review]
+1

**Pull request merged**